### ISSUE TITLE
`should contain only` is broken

### ISSUE THREAD

**Issue by mr-calvert** [scalatest/scalatest#492]
Consider:
`Vector(1) should contain only (1, 2, 3)`
fails while
`Vector(2, 3) should contain only (1, 2, 3)
succeeds.

I dug a little and found commit https://github.com/scalatest/scalatest/commit/5e1f9baf17f36328ef7e34d14c62172ea2b7c0b9 which seems to have attempted to change the semantics of Aggregating.checkOnly to mean:
(lhs ⊂ rhs) and (rhs ⊂ lhs)
but as I noted on that commit, that's not what it does.

The commit in question checks that lhs is one of rhs's tails, and is very dependent on the order of rhs. Sorry to be all ranty, but I don't agree with the design intent of that commit. I'd rather `x should contain only y` mean:
lhs ⊂ rhs
and let
`x should contain theSameElementsAs y` do stand for the use case the commit was trying to address.

Thanks,
Scott Calvert

**Comment by bvenners** [comment]
No worries. We really appreciate the feedback. We'll revisit the actual intent, clarify it, and make sure the code is correct.

Bill

**Comment by mr-calvert** [comment]
I'm happy to submit a PR if a core team member can say what the intended semantics for `x should contain only y` are.

**Comment by mr-calvert** [comment]
Thanks, I'm happy to PR the code change once the intent is clarified.

**Comment by bvenners** [comment]
Thanks. We have put this on our queue, but are in the middle of things this week prepping for nescala Friday. So might be next week before we can get sufficient quiet time to study the intended semantics. Will comment here once we've done that.

Bill

**Comment by mr-calvert** [comment]
Thanks, no need to hurry on my account. I'll be done refactoring our tests away from `only` before the end of the day.

**Comment by bvenners** [comment]
We are going to make another attempt at the bug fix, but I thought about it and decided to continue with the same symantics. My original idea of "only" was that it means the same as "allOf" but doesn't allow any extras in the left hand side. Also, to get the looser semantics that you prefer you can do this:

all (Vector(1)) should (equal (1) or equal (2) or equal (3))

**Comment by franckrasolo** [comment]
I too was recently surprised by the following:
```
class WeirdSpec extends WordSpec with MustMatchers {
  "fails as expected" in {
    Seq(1) must contain only (1, 2)
  }

  "passes unexpectedly" in {
    Seq(1) must contain only (2, 1)
  }
}
```

Surely the second test ought to fail, regardless of semantics. Shouldn't it?

**Comment by bvenners** [comment]
Yes, it was a bug, unfortunately. We have fixed it already in 2.3.x. Will be released in 2.3.0.

scala> :paste
// Entering paste mode (ctrl-D to finish)

class WeirdSpec extends WordSpec with MustMatchers {
  "fails as expected" in {
    Seq(1) must contain only (1, 2)
  }

  "passes unexpectedly" in {
    Seq(1) must contain only (2, 1)
  }
}

// Exiting paste mode, now interpreting.

defined class WeirdSpec

scala> run(new WeirdSpec)
WeirdSpec:
- fails as expected *** FAILED ***
  List(1) did not contain only (1, 2) (<console>:18)
- passes unexpectedly *** FAILED ***
  List(1) did not contain only (2, 1) (<console>:22)

**Comment by franckrasolo** [comment]
That's great to hear, thanks Bill!

**Comment by franckrasolo** [comment]
Any idea when some sort of preview of 2.3.0 will be available?

**Comment by Vlad-Sadovsky** [comment]
Empty Seq 'contains only' items from any other Seq.
The only exclusion is Type mismatch, that can be checked at compile time.
`
import org.scalatest.{MustMatchers, WordSpec}

class ContainsOnlyTest extends WordSpec with MustMatchers {
  "Empty Seq must contain only (1, 2)" in {
    Seq[Int]() must contain only (1, 2)
  }
}`
Must pass but it failed for scalatest_2.12-3.0.8.jar
Failed with "List() did not contain only (1, 2)"
